src/types/poly_func.rs

```diff
@@ -0,0 +1,524 @@
+//! Polymorphic Function Types
+
+use crate::extension::{ExtensionRegistry, SignatureError};
+use itertools::Itertools;
+
+use super::{
+    type_param::{check_type_args, TypeArg, TypeParam},
+    FunctionType, Substitution,
+};
+
+/// A polymorphic function type, e.g. of a [Graph], or perhaps an [OpDef].
+/// (Nodes/operations in the Hugr are not polymorphic.)
+///
+/// [Graph]: crate::values::PrimValue::Function
+/// [OpDef]: crate::extension::OpDef
+#[derive(
+    Clone, PartialEq, Debug, Eq, derive_more::Display, serde::Serialize, serde::Deserialize,
+)]
+#[display(
+    fmt = "forall {}. {}",
+    "params.iter().map(ToString::to_string).join(\" \")",
+    "body"
+)]
+pub struct PolyFuncType {
+    /// The declared type parameters, i.e., these must be instantiated with
+    /// the same number of [TypeArg]s before the function can be called.
+    ///
+    /// [TypeArg]: super::type_param::TypeArg
+    pub(crate) params: Vec<TypeParam>,
+    /// Template for the function. May contain variables up to length of [Self::params]
+    pub(super) body: FunctionType,
+}
+
+impl From<FunctionType> for PolyFuncType {
+    fn from(body: FunctionType) -> Self {
+        Self {
+            params: vec![],
+            body,
+        }
+    }
+}
+
+impl PolyFuncType {
+    /// Create a new PolyFuncType and validates it. (This will only succeed
+    /// for outermost PolyFuncTypes i.e. with no free type-variables.)
+    /// The [ExtensionRegistry] should be the same (or a subset) of that which will later
+    /// be used to validate the Hugr; at this point we only need the types.
+    ///
+    /// #Errors
+    /// Validates that all types in the schema are well-formed and all variables in the body
+    /// are declared with [TypeParam]s that guarantee they will fit.
+    pub fn new_validated(
+        params: impl Into<Vec<TypeParam>>,
+        body: FunctionType,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<Self, SignatureError> {
+        let params = params.into();
+        body.validate(extension_registry, &params)?;
+        Ok(Self { params, body })
+    }
+
+    pub(super) fn validate(
+        &self,
+        reg: &ExtensionRegistry,
+        external_type_vars: &[TypeParam],
+    ) -> Result<(), SignatureError> {
+        // TODO should we add a mechanism to validate a TypeParam?
+        let mut v;
```

Comment:
Yes!!! I followed `rustc` and `clippy`s direction. The point of declaring here is so that the value in it lives until the end of this scope, rather than the much-shorter else scope, so we can borrow said value until end of this scope

Reply:
Added a comment